File: Data_Streaming_Processing/Q1. Kth Largest Element in a Stream/Q1_KthLargestElementinaStream.java
import java.util.PriorityQueue;      // import min-heap structure

class KthLargest {

    private PriorityQueue<Integer> minHeap; // heap to store k largest numbers
    private int k;                          // which kth largest we need

    public KthLargest(int k, int[] nums) {
        this.k = k;                         // store k
        minHeap = new PriorityQueue<>();    // create an empty min-heap

        for (int n : nums) {                // loop through initial numbers
            minHeap.offer(n);               // add number to heap
            if (minHeap.size() > k) {       // if heap has more than k numbers
                minHeap.poll();             // remove smallest number
            }
        }
    }
    
    public int add(int val) {
        minHeap.offer(val);                 // add new value to heap
        
        if (minHeap.size() > k) {           // if heap is larger than k
            minHeap.poll();                 // remove smallest number
        }
        
        return minHeap.peek();              // return kth largest (top of heap)
    }
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest obj = new KthLargest(k, nums);
 * int param_1 = obj.add(val);
 */

File: Data_Streaming_Processing/Q1. Kth Largest Element in a Stream/Q1_KthLargestElementinaStream.py
import heapq

class KthLargest(object):

    def __init__(self, k, nums):
        """
        :type k: int
        :type nums: List[int]
        """
        self.k = k                      # store k
        self.min_heap = []              # create empty min-heap
        
        for n in nums:                  # add all initial numbers
            heapq.heappush(self.min_heap, n)   # push into heap
            if len(self.min_heap) > k:         # if heap too big
                heapq.heappop(self.min_heap)   # remove smallest

    def add(self, val):
        """
        :type val: int
        :rtype: int
        """
        heapq.heappush(self.min_heap, val)     # add new value
        
        if len(self.min_heap) > self.k:        # keep only k elements
            heapq.heappop(self.min_heap)       # remove smallest
        
        return self.min_heap[0]                # kth largest is heap root


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)
